thrust/thrust/system/hpx/detail/find.h

```diff
@@ -26,5 +30,81 @@
 #  pragma system_header
 #endif // no system header
 
-// this system inherits find
-#include <thrust/system/cpp/detail/find.h>
+#include <thrust/iterator/iterator_traits.h>
+#include <thrust/system/hpx/detail/execution_policy.h>
+#include <thrust/system/hpx/detail/runtime.h>
+
+#include <hpx/parallel/algorithms/find.hpp>
+
+THRUST_NAMESPACE_BEGIN
+namespace system
+{
+namespace hpx
+{
+namespace detail
+{
+
+// Helper trait to detect if an iterator's reference type can be bound to a non-const lvalue reference
+// HPX's parallel algorithms expect this property, but many Thrust iterators return rvalues
+template <typename Iterator>
+struct iterator_reference_is_lvalue_reference
+{
+  using reference_type        = typename thrust::iterator_traits<Iterator>::reference;
+  static constexpr bool value = std::is_lvalue_reference_v<reference_type>;
+};
+
+template <typename DerivedPolicy, typename InputIterator, typename T>
+InputIterator find(execution_policy<DerivedPolicy>& exec, InputIterator first, InputIterator last, const T& value)
+{
+  if constexpr (::hpx::traits::is_forward_iterator_v<InputIterator>
+                && iterator_reference_is_lvalue_reference<InputIterator>::value)
+  {
+    return hpx::detail::run_as_hpx_thread([&] {
+      return ::hpx::find(hpx::detail::to_hpx_execution_policy(exec), first, last, value);
+    });
+  }
+  else
+  {
+    (void) exec;
+    return ::hpx::find(first, last, value);
+  }
+}
+
+template <typename DerivedPolicy, typename InputIterator, typename Predicate>
+InputIterator find_if(execution_policy<DerivedPolicy>& exec, InputIterator first, InputIterator last, Predicate pred)
```

Comment:
Wrap `pred` so that we don't need to rely on implicit transformations.

See the implementation of [reduce](https://github.com/STEllAR-GROUP/cccl/blob/837e793a265aa3b2a4faa40b8ac64e93051e4a72/thrust/thrust/system/hpx/detail/reduce.h#L54).